### ISSUE TITLE
Unable to run Jekyll server locally

### ISSUE THREAD

**Issue by noah-weingarden** [eecs485staff/primer-spec#191]
**Describe the bug**
When I attempt to run the Jekyll server for the demo in this repo, I receive the following error:

```console
[2022-07-06 22:19:21] ERROR Errno::ECONNRESET: Connection reset by peer @ io_fillbuf - fd:15 
	/home/noah/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/webrick-1.7.0/lib/webrick/httpserver.rb:82:in `eof?'
	/home/noah/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/webrick-1.7.0/lib/webrick/httpserver.rb:82:in `run'
	/home/noah/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/webrick-1.7.0/lib/webrick/server.rb:310:in `block in start_thread'
```
The error appears each time I refresh the page in my browser. Consequently, the rendered DOM doesn't contain any visible elements, leading to a blank screen. This happens the majority of the time I refresh the browser after I run `./script/server` or manually run `bundle exec jekyll serve`. Sometimes the error will randomly not occur and the page will render correctly.

**To Reproduce**
1. Create a new branch off of `develop` using a freshly cloned repo.
2. Run `./script/bootstrap`.
3. Run `./script/server`.
4. Visit `localhost:4000` in the browser.
5. See error.

**Expected behavior**
I expect no errors and the page's content to be rendered properly.

**Desktop (please complete the following information):**
 - OS: Ubuntu 20.04
 - Browser: Chrome 101
 - Primer version: 1.7.1
 - Ruby version: 2.7.4
 - Node version: 16.15.1

**Additional context**
I've never observed this error while previewing Primer Spec for an EECS 485 project.

**Comment by seshrs** [comment]
Hi @noah-weingarden, thanks for trying to build Primer Spec locally! Also thanks for the detailed issue report :)

That's an interesting error, I haven't encountered that before. It might be some incompatibility between Jekyll and Webrick. (Webrick is a dependency of Jekyll, it powers the dev server.) I see several posts about this error online, but I don't see definitive solutions.

A couple of things to investigate:
- What is the full error message? Are there any warnings just above the `Connection reset by peer` error that you shared?
- What version of Webrick are you using? The latest version is 1.7.0. (Run `bundle info webrick` to find the version.)
  - If your version is out of date, what happens when you run `bundle update webrick`?
- Similarly, what version of Jekyll are you using? (I'm currently running 3.9.0.)

**Comment by noah-weingarden** [comment]
> thanks for trying to build Primer Spec locally! Also thanks for the detailed issue report :)

Of course! :)

> What is the full error message? Are there any warnings just above the Connection reset by peer error that you shared?

No, that's the entire error. I also get the standard expected output. The full output is this:

```console
Configuration file: /home/noah/primer-spec/_config.yml
            Source: /home/noah/primer-spec
       Destination: /home/noah/primer-spec/_site
 Incremental build: disabled. Enable with --incremental
      Generating... 
      Remote Theme: Using theme eecs485staff/primer-spec
To use retry middleware with Faraday v2.0+, install `faraday-retry` gem
                    done in 3.118 seconds.
 Auto-regeneration: enabled for '/home/noah/primer-spec'
    Server address: http://127.0.0.1:4000
  Server running... press ctrl-c to stop.
[2022-07-06 22:55:15] ERROR Errno::ECONNRESET: Connection reset by peer @ io_fillbuf - fd:15 
	/home/noah/.rbenv/versions/2.7.4/lib/ruby/2.7.0/webrick/httpserver.rb:82:in `eof?'
	/home/noah/.rbenv/versions/2.7.4/lib/ruby/2.7.0/webrick/httpserver.rb:82:in `run'
	/home/noah/.rbenv/versions/2.7.4/lib/ruby/2.7.0/webrick/server.rb:307:in `block in start_thread'
```

The `To use retry middleware with Faraday v2.0+, install faraday-retry gem` part might be kind of unusual too though.

> What version of Webrick are you using? The latest version is 1.7.0. (Run bundle info webrick to find the version.)

I'm running version 1.7.0.

> Similarly, what version of Jekyll are you using? (I'm currently running 3.9.0.)

I was running 3.9.2 before. I just ran `gem install jekyll`, and it supposedly installed 4.2.2, but when I run `bundle info jekyll`, it still tells me I have 3.9.2.

I also tried googling about this, but I couldn't find anything helpful. :/ Deleting `_site` didn't work, and there doesn't appear to be a `.jekyll_cache`. I also tried replacing webrick with thin (since some online posts suggested that), and it did succeed at preventing any errors, but it didn't solve the problem, as nothing was rendered.

**Comment by seshrs** [comment]
> The `To use retry middleware with Faraday v2.0+, install faraday-retry gem` part might be kind of unusual too though.

That _does_ look like it might be the fix. Could you try following these steps:
1. Add a line to `Gemfile` underneath the `webrick` line: `gem faraday-retry`
2. Run: `bundle install`.
3. Verify that it was installed correctly using `bundle info faraday-retry`. (I’m AFK so I can’t try this yet, but sometimes you need to run `bundle update` also.)
4. Try running `bundle exec jekyll serve` to verify if the issue repro’s.

---

Since webrick is a dependency of Jekyll, I don’t think we can replace it with thin. 

Also, the version of gems that you have installed on your Ruby environment may not correspond with the version that bundle has installed for a particular project. (Bundler is kind-of like Python virtual environments in that it isolates your project to use the gems that will work.)
Primer Spec is designed to work with GitHub Pages, [which will not use Jekyll 4 for the foreseeable future](https://github.com/github/pages-gem/issues/651#issuecomment-581069671). That’s why I’ve constrained the Jekyll version to make sure we don’t use unsupported Jekyll features!

Going forward: I think I should commit my `Gemfile.lock` file, just like in the project spec repos. I wonder if that might have also fixed your issue.

**Comment by noah-weingarden** [comment]
No good--I've verified that I have `faraday-retry` version 2.0.0, but the error still appears.

> Also, the version of gems that you have installed on your Ruby environment may not correspond with the version that bundle has installed for a particular project. (Bundler is kind-of like Python virtual environments in that it isolates your project to use the gems that will work.)
Primer Spec is designed to work with GitHub Pages, https://github.com/github/pages-gem/issues/651#issuecomment-581069671. That’s why I’ve constrained the Jekyll version to make sure we don’t use unsupported Jekyll features!

Ah, thanks for the explanation!

> Going forward: I think I should commit my Gemfile.lock file, just like in the project spec repos. I wonder if that might have also fixed your issue.

That's worth a try once you're able to commit it!

**Comment by seshrs** [comment]
Closed by #198.
